File: examples/test-echo/processor.py
#!/usr/bin/env python

import sys, json, datetime, string, os, pytz

def time_format(dt):
    return "%s:%.3f%s" % (
        dt.strftime('%Y-%m-%dT%H:%M'),
        float("%.3f" % (dt.second + dt.microsecond / 1e6)),
        dt.strftime('%z')
    )

while True:
    try:
        # Assings ENV variables to an object
        obj = os.environ
        # Reads JSON from stdin
        obj["PAYLOAD"] = sys.stdin.readline().replace("\n", "")

        if obj["PAYLOAD"] != "":
            # first we check if the payload is valid JSON
            json_payload = json.loads(obj["PAYLOAD"])
            # Generates datetime ISO 8601 format ex. 2017-04-24T21:16:26.678Z
            obj["CREATED_AT"] = time_format(datetime.datetime.now(pytz.utc)).replace('+0000', 'Z')
            # Open JSON template
            f = open('./service-processor/schema-test-echo.json', 'r')
            lines = f.readlines()
            # Strips JSON and then output a single line string
            json_str = ''.join([line.strip() for line in lines])
            # Replace ENV variables in the JSON template with values and print to stdout
            # Must use print and not sys.stdout.write otherwise it will not work
            print(string.Template(json_str).substitute(obj))
            sys.stdout.flush()

    except ValueError:
        # If JSON from stdin is invalid
        sys.stdout.write('{"topic": "' + obj["NAMESPACE"] + '/log", "message": "http-request: error"}' + '\n')
        sys.stdout.flush()

File: examples/test-echo/processor.js
var readline = require('readline');
var fs = require('fs');
var expandenv = require('expandenv');

// Reads JSON from stdin
var rl_stdin = readline.createInterface({
    input: process.stdin
});
// When line comes from stdin
rl_stdin.on('line', function(line){
    try {
        // Assings ENV variables to an object
        obj = process.env;
        if (line !== null) {
            // first we check if the payload is valid JSON
            json_payload = JSON.parse(line);
            obj.PAYLOAD = line.trim().toString();
            // Generates datetime
            obj.CREATED_AT = new Date().toISOString();
            // Open JSON template
            var rl_template = readline.createInterface({
                input: fs.createReadStream('./service-processor/schema-test-echo.json')
            });
            var result = '';
            // concat lines from template
            rl_template.on('line', function(line) {
                if (line !== null) {
                    // Trims JSON and then output a single line string
                    result += line.trim().toString();
                }
            });
            // All lines read and concatenated
            rl_template.on('close', () => {
                // Replace ENV variables in the JSON template with values and print to stdout
                console.log(expandenv(result));
            });
        }
    } catch (e) {
        console.log('{"topic": "' + process.env.NAMESPACE + '/log", "message": "http-request: error"}');
    }
});
